File: Parcial/Resolucion C#/Resolucion Parcial C#/Form1.cs
using Services.Model;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Windows.Forms.VisualStyles;

namespace Resolucion_Parcial_C_
{
    public partial class Form1 : Form
    {
        #region Dependencias
        private DataTable dtPinturas;
        #endregion

        #region Inicializar
        private void InicializarDataTable()
        {
            DataTable dt = new DataTable();

            dt.Columns.Add("CODIGO", typeof(long));
            dt.Columns.Add("COLOR", typeof(string));
            dt.Columns.Add("LITROS", typeof(double));

            this.SetDataTable(dt);
        }

        #endregion

        #region Constructor
        public Form1()
        {
            InitializeComponent();
        }

        #endregion

        #region Acciones

        // ---------------------- Seteo la DATA TABLE con el DATA GRID ---------------------- //
        private void SetDataTable(DataTable dt)
        {
            this.dtPinturas = dt;
            dtGrid_Pinturas.DataSource = dt;
        }

        // ---------------------- Funcion calcualar el total de listros pintura roja ---------------------- //
        private void CalcularTotalLitrosPinturaRoja()
        {

            double acumuladorLitrosRojo = 0;

            foreach (DataRow renglon in this.dtPinturas.Rows)
            {
                string color = renglon["COLOR"].ToString();

                if (color.ToUpper() == "ROJO" || color.ToUpper() == "ROJA")
                {
                    acumuladorLitrosRojo += Convert.ToDouble(renglon["LITROS"]);
                }
            }

            lbl_totalRojo.Text = acumuladorLitrosRojo.ToString("#0.00");

        }

        // ---------------------- Funcion calcualar el promedio de listros de pintura azul ---------------------- //
        private void CalcularPromedioLitrosPinturaAzul()
        {

            int contadorPinturaAzul = 0;
            double acumuladorLitrosAzul = 0;

            foreach (DataRow renglon in this.dtPinturas.Rows)
            {
                string color = renglon["COLOR"].ToString();

                if (color.ToUpper() == "AZUL")
                {
                    acumuladorLitrosAzul += Convert.ToDouble(renglon["LITROS"]);
                    contadorPinturaAzul++;
                }
            }

            double promedio = acumuladorLitrosAzul / contadorPinturaAzul;

            if (promedio > 0)
            {
                lbl_promedioAzul.Text = promedio.ToString("#0.00");
            }
            else
            {
                lbl_promedioAzul.Text = "0.00";
            }
        }

        // ---------------------- Funcion limpiar los TextBox ---------------------- //
        private void LimpiarTextBox()
        {
            txt_Codigo.Text = "";
            txt_Color.Text = "";
            txt_Litros.Text = "";
        }
        #endregion

        #region Eventos

        // ---------------------- Evento LOAD ---------------------- //
        private void Form1_Load(object sender, EventArgs e)
        {
            this.InicializarDataTable();
        }

        // ---------------------- Evento CLICK para agregar pintura ---------------------- //
        private void btn_AgregarPintura_Click(object sender, EventArgs e)
        {
            // Primero tomo los valores de los TextBoxes
            string codigoS = txt_Codigo.Text;
            string color = txt_Color.Text;
            string litrosS = txt_Litros.Text;

            // PRIMERA VALIDACION: QUE NO HAYA CAMPOS VACIOS
            if (codigoS.Equals("") || color.Equals("") || litrosS.Equals(""))
            {
                MessageBox.Show("ERROR: Algunos campos estan vacios");
                return;
            }

            // SEGUNDA VALIDACION: QUE CODIGO Y LITROS SEAN DATOS NUMERICOS
            int codigo;
            double litros;

            if (!int.TryParse(codigoS, out codigo) || !double.TryParse(litrosS, out litros))
            {
                MessageBox.Show("ERROR: Los campos CODIGO y LITROS deben ser numericos");
                return;
            }

            // TERCERA VALIDACION: QUE LOS VALORES DE CODIGO Y LISTROS NO SEAN NEGATIVOS
            if (codigo < 0 || litros < 0)
            {
                MessageBox.Show("ERROR: Numero negativo, por favor ingrese valores POSITIVOS");
                return;
            }

            // CUARTA VALIDACION QUE CODIGO SEA UNICO
            foreach (DataRow renglon in dtPinturas.Rows)
            {
                if (renglon["CODIGO"].ToString() == codigo.ToString())
                {
                    MessageBox.Show("ERROR: Codigo repetido, por favor ingrese uno diferente");
                    return;
                }
            }

            // Si pasaron todas las validaciones creo un nuevo objeto de tipo Pintura
            Pintura nuevaPintura = new Pintura(codigo, color, litros);

            // Agregamos una FILA conlas propiedades del nuevo objeto
            this.dtPinturas.Rows.Add(nuevaPintura.codigo, nuevaPintura.color, nuevaPintura.litros);

            // Ejecuto las funciones de mi region Acciones
            this.CalcularTotalLitrosPinturaRoja();
            this.CalcularPromedioLitrosPinturaAzul();
            this.LimpiarTextBox();
        }

    }
    #endregion




}

File: Parcial/Resolucion VB/Resolucion Parcial/Form1.vb
Public Class Form1

#Region "Dependencias"
    '----------------------- LA DATATABLE CON LA QUE VAMOS A TRABAJR -----------------------'
    Private dtPinturas As New DataTable
#End Region

#Region "Inicializar"

    '----------------------- FUNCION PARA CREAR LA DATATABLE -----------------------'
    Private Sub InicializarDataTable()
        Dim dt As New DataTable

        dt.Columns.Add("CODIGO")
        dt.Columns.Add("COLOR")
        dt.Columns.Add("LITROS")

        Me.SetDataTable(dt)
    End Sub
#End Region

#Region "Acciones"

    '----------------------- FUNCION PARA SETEAR LA DATATABLE CON LA QUE VAMOS A TRABAJAR CON EL DATAGRID -----------------------'
    Private Sub SetDataTable(dt As DataTable)
        Me.dtPinturas = dt
        dtgrid_Pinturas.DataSource = dt
    End Sub

    '----------------------- FUNCION CALCULAR TOTAL LITROS ROJO -----------------------'
    Private Sub CalcularTotalLitrosPinturaRoja()
        Dim acumuladorLitrosRojo As Double = 0

        For Each renglon As DataRow In Me.dtPinturas.Rows
            Dim color As String = renglon("COLOR")

            If color.ToUpper() = "ROJO" Or color.ToUpper() = "ROJA" Then
                acumuladorLitrosRojo += renglon("LITROS")
            End If
        Next

        lbl_litrosRojo.Text = acumuladorLitrosRojo.ToString("#0.00")
    End Sub

    '----------------------- FUNCION CALCULAR PROMEDIO LITROS AZUL -----------------------'
    Private Sub CalcularPromedioLitrosPinturaAzul()
        Dim acumuladorLitrosAzul As Double = 0
        Dim contadorPinturaAzul As Integer = 0

        For Each renglon As DataRow In Me.dtPinturas.Rows
            Dim color As String = renglon("COLOR")

            If color.ToUpper() = "AZUL" Then
                acumuladorLitrosAzul += renglon("LITROS")
                contadorPinturaAzul += 1
            End If
        Next

        Dim promedioLitrosAzul As Double = acumuladorLitrosAzul / contadorPinturaAzul

        If contadorPinturaAzul > 0 Then
            lbl_litrosAzul.Text = promedioLitrosAzul.ToString("#0.00")
        Else
            lbl_litrosAzul.Text = "0.00"
        End If


    End Sub

    '----------------------- FUNCION LIMIPIAR TEXTBOX -----------------------'
    Private Sub LimparTextBox()
        txt_codigo.Text = ""
        txt_color.Text = ""
        txt_litros.Text = ""
    End Sub
#End Region

#Region "Eventos"

    '----------------------- EVENTO LOAD CARGAR FORMULARIO -----------------------'
    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Me.InicializarDataTable()
    End Sub

    '----------------------- EVENTO CLICK AGREGAR PINTURA -----------------------'
    Private Sub btn_agregarPintura_Click(sender As Object, e As EventArgs) Handles btn_agregarPintura.Click

        'Primero tomo los valores de los TextBoxes'
        Dim codigoS = txt_codigo.Text
        Dim color = txt_color.Text
        Dim litrosS = txt_litros.Text

        'PRIMERA VALIDACION: QUE NO HAYA CAMPOS VACIOS'
        If codigoS = "" Or color = "" Or litrosS = "" Then
            MessageBox.Show("ERROR: Algunos campos estan vacios")
            Return
        End If

        'SEGUNDA VALIDACION: QUE CODIGO Y LITROS SEAN VALORES NUMERICOS'
        Dim codigo As Integer
        Dim litros As Double

        If Not Integer.TryParse(codigoS, codigo) Or Not Double.TryParse(litrosS, litros) Then
            MessageBox.Show("ERROR: Los valores de CODIGO o LITROS deben ser NUMÉRICOS")
            Return
        End If

        'TERCERA VALIDACION: QUE CODIGO Y LITROS SEAN POSITIVOS'
        If codigo < 0 Or litros < 0 Then
            MessageBox.Show("ERROR: Número negativo, por favor ingrese valores POSITIVOS")
            Return
        End If

        'CUARTA VALIDACION: QUE NO HAYA CODIGO REPETIDO'
        For Each renglon As DataRow In Me.dtPinturas.Rows
            If renglon("CODIGO") = codigo Then
                MessageBox.Show("ERROR: Código repetido, por favor ingrese uno diferente")
                Return
            End If
        Next

        'Si pasaron todas las validaciones agrego una nueva fila a la DataTable con los valores de los TextBox'
        Me.dtPinturas.Rows.Add(codigo, color, litros)

        'Ejecuto las funciones dem i region Acciones'
        Me.CalcularTotalLitrosPinturaRoja()
        Me.CalcularPromedioLitrosPinturaAzul()
        Me.LimparTextBox()

    End Sub
#End Region


End Class
